c/tree.c

```diff
@@ -672,56 +797,100 @@ LOCAL MMDBW_status insert_record_into_next_node(
+        // locations. If that happens, we end up freeing the node multiple
+        // times when destroying the tree.
+        //
+        // Consider the case where we previously inserted fixed nodes to a
+        // greater depth than we are right now. Because we'd continue to
+        // descend the tree, we'd insert this new fixed node record at
+        // potentially multiple spots in the tree, at each point we encounter
+        // an EMPTY, DATA, or NODE record (the conditions where we can enter
+        // insert_record_into_current_record()). The stopping condition would
+        // not be correct.
+        return MMDBW_FIXED_NODE_OVERWRITE_ATTEMPT_ERROR;
+    }
+
+    // Figure out the next node.
     MMDBW_node_s *next_node = NULL;
     switch (current_record->type) {
     case MMDBW_RECORD_TYPE_EMPTY:
     case MMDBW_RECORD_TYPE_DATA: {
+            // In this case we create a new node to point to. We make the new
+            // nodes left and right identical to us.
             next_node = new_node_from_record(tree, current_record);
             current_record->value.node = next_node;
             current_record->type = MMDBW_RECORD_TYPE_NODE;
             break;
         }
+    case MMDBW_RECORD_TYPE_FIXED_EMPTY:
     case MMDBW_RECORD_TYPE_ALIAS: {
-            // The insert is trying to overwrite an aliased network. We do not
-            // allow this.
+            // The insert is trying to overwrite an aliased network or one set
+            // fixed empty. We do not allow this.
             if (current_bit == network->prefix_length) {
-                return MMDBW_ALIAS_OVERWRITE_ATTEMPT_ERROR;
+                return current_record->type == MMDBW_RECORD_TYPE_FIXED_EMPTY ?
+                       MMDBW_FIXED_EMPTY_OVERWRITE_ATTEMPT_ERROR :
+                       MMDBW_ALIAS_OVERWRITE_ATTEMPT_ERROR;
             }
-            // We don't follow aliases when inserting a network that contains an
-            // aliased network within it.
-            else if (current_bit > network->prefix_length) {
+            // We don't follow aliases when inserting a network that contains
+            // an aliased network within it. Nor do we accept networks
+            // containing networks present as fixed empty records.
+            if (current_bit > network->prefix_length) {
+                // We return success to silently ignore when we try to insert
+                // here. If we raise an error, it makes working with dirty data
+                // more difficult. We assume such inserts were not intended and
+                // can be safely skipped.
                 return MMDBW_SUCCESS;
-            } else {
-                return MMDBW_INSERT_INTO_ALIAS_NODE_ERROR;
             }
+            // current_bit < network->prefix_length. This means we contain the
+            // new network already as FIXED_EMPTY/ALIAS. Inserting the network
+            // is not valid because of that.
```

Comment:
One concern I have about making this behave the same was as alias nodes--which I realize I suggested--is that it is a fairly significant behavior change and we will now throw an exception when inserting into a private network. We will have to add checks to the GeoIP2 build to ensure that we don't do this. Another option that is more similar to the old behavior is to just silently drop the data. I am not sure I love that either though.

Reply:
Yeah that's a good point. We do silently drop the inserts already if they contain the network, as you can see, so doing so for into/replacement would probably be okay and would not be too difficult. It would add some consistency, arguably.

I wonder if we should add an option to be strict vs not. But I think we should probably not complicate this code by adding toggles if possible.

One argument in favour of having errors might be that it could lead to cleaning up data by the callers, but I don't know how much work that would generate, or if it would be even valuable.

An argument in favour of being silent is that was the prior behaviour!

I'm open to either way. Not sure which to go with. We could perhaps raise it with someone involved in the build and see what they think.